lib/processors/manifestTransformer.js

```diff
@@ -0,0 +1,218 @@
+			bundleUrl: sapAppBundleConfig
+		};
+	}
+
+	if (sapAppBundleConfig) {
+		sapAppConfig.i18n = sapAppBundleConfig;
+		bundleConfigs.push(sapAppBundleConfig);
+	}
+	return bundleConfigs;
+}
+
+function getSapUi5ModelBundles(manifest, bundleConfigs) {
+	const sapui5ModelConfig = manifest?.["sap.ui5"]?.models || {};
+	return bundleConfigs.concat(Object.values(sapui5ModelConfig)
+		.filter((model) => model.type === "sap.ui.model.resource.ResourceModel")
+		.map((model) => model.settings));
+}
+
+function getSapUi5LibrariesBundles(manifest, bundleConfigs) {
+	const sapui5LibraryConfig = manifest?.["sap.ui5"]?.library;
+	let sapui5LibraryBundleConfig = sapui5LibraryConfig?.i18n;
+
+	if (!sapui5LibraryBundleConfig) {
+		return bundleConfigs;
+	}
+
+	if (typeof sapui5LibraryBundleConfig === "string") {
+		sapui5LibraryBundleConfig = {
+			bundleUrl: sapui5LibraryBundleConfig
+		};
+	} else if (typeof sapui5LibraryBundleConfig === "boolean") {
+		sapui5LibraryBundleConfig = {
+			bundleUrl: "messagebundle.properties"
+		};
+	}
+
+	if (sapui5LibraryBundleConfig) {
+		sapui5LibraryConfig.i18n = sapui5LibraryBundleConfig;
+		bundleConfigs.push(sapui5LibraryBundleConfig);
+	}
+	return bundleConfigs;
+}
+
+function getBundles(manifest) {
+	let bundleConfigs = [];
+	const appType = getType(manifest);
+
+	bundleConfigs = getSapAppBundle(manifest, bundleConfigs);
+
+	if (appType === "library") {
+		bundleConfigs = getSapUi5LibrariesBundles(manifest, bundleConfigs);
+	} else {
+		bundleConfigs = getSapUi5ModelBundles(manifest, bundleConfigs);
+	}
+
+	return bundleConfigs
+		.filter((bundleConfig) => isAutofillRequired(bundleConfig, manifest["sap.app"].id));
+}
+
+async function transformManifest(resource, fs, options) {
+	// merge options with defaults
+	options = Object.assign({
+		prettyPrint: true,
+	}, options);
+	const content = await resource.getString();
+	const manifest = JSON.parse(content);
+
+	if (!manifest._version) {
+		log.verbose("manifest.json: version is not defined. No supportedLocales are generated");
+		return;
+	}
+
+	const descriptorVersion = new Version(manifest._version);
+	if (descriptorVersion.compare(APP_DESCRIPTOR_V22) === -1) {
+		log.verbose("manifest.json: version is lower than 1.21.0 so no supportedLocales can be generated");
+		return;
+	}
+
+	const bundles = getBundles(manifest);
+
+	if (bundles.length === 0) {
+		return;
+	}
+
+	// check which locales are avalable
+	await Promise.all(bundles.map(async (bundleConfig) => {
+		const {bundleUrl, bundleName, fallbackLocale} = bundleConfig;
+		let propertyFilesPath = bundleUrl;
+
+		if (bundleName) {
+			propertyFilesPath = bundleName.replace(/\./g, "/") + ".properties";
+		}
+
+		const propertyFileName = propertyFilesPath.substring(propertyFilesPath.lastIndexOf("/") + 1);
+		propertyFilesPath = propertyFilesPath.substring(0, propertyFilesPath.lastIndexOf("/"));
+
+		const generatedSupportedLocales = fs.readdir(`/resources/${propertyFilesPath}/!(${propertyFileName}`)
```

Comment:
fs.readdir is async and accepts a callback. It's the default Node.js "fs" interface. It usually makes sense to use util.promisify to get a promise version of a function.